crypto/keys/pem_signer_factory.go

```diff
@@ -41,9 +47,49 @@ func (f PEMSignerFactory) NewSigner(ctx context.Context, tree *trillian.Tree) (c
 	}
 
 	switch privateKey := privateKey.Message.(type) {
-	case *trillian.PEMKeyFile:
+	case *keyspb.PEMKeyFile:
 		return NewFromPrivatePEMFile(privateKey.GetPath(), privateKey.GetPassword())
+	case *keyspb.PrivateKey:
+		return NewFromPrivateDER(privateKey.GetDer())
 	}
 
 	return nil, fmt.Errorf("unsupported PrivateKey type for tree %d: %T", tree.GetTreeId(), privateKey.Message)
 }
+
+// Generate creates a new private key for a tree based on a key specification.
+// It returns a proto that can be used as the value of tree.PrivateKey.
+func (f PEMSignerFactory) Generate(ctx context.Context, tree *trillian.Tree, spec *keyspb.Specification) (*any.Any, error) {
+	if tree.PrivateKey != nil {
+		return nil, fmt.Errorf("tree already has a private key")
+	}
+
+	key, err := NewFromSpec(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key: %v", err)
+	}
+
+	if SignatureAlgorithm(key.Public()) != tree.GetSignatureAlgorithm() {
+		return nil, fmt.Errorf("Specification produces %T, but tree.SignatureAlgorithm = %v", key, tree.GetSignatureAlgorithm())
+	}
+
+	return marshalPrivateKey(key)
+}
+
+func marshalPrivateKey(key crypto.Signer) (*any.Any, error) {
+	var privateKeyDER []byte
+	var err error
+
+	switch key := key.(type) {
+	case *ecdsa.PrivateKey:
+		privateKeyDER, err = x509.MarshalECPrivateKey(key)
+	case *rsa.PrivateKey:
+		privateKeyDER = x509.MarshalPKCS1PrivateKey(key)
+	default:
+		return nil, fmt.Errorf("unsupported key type: %T", key)
+	}
+	if err != nil {
```

Comment:
nit: Move `var err error`  and the if below line 84? That's the only branch with errors.